keras_nlp/layers/relative_attention.py

```diff
@@ -0,0 +1,560 @@
+    equation = "%s,%s->%s" % (input_str, kernel_str, output_str)
+
+    return equation, bias_axes, len(output_str)
+
+
+def _get_output_shape(output_rank, known_last_dims):
+    return [None] * (output_rank - len(known_last_dims)) + list(known_last_dims)
+
+
+def _rel_shift(x, klen=-1):
+    """Performs relative shift to form the relative attention score."""
+
+    x = tf.transpose(x, perm=[2, 3, 0, 1])
+    x_size = tf.shape(x)
+
+    x = tf.reshape(x, [x_size[1], x_size[0], x_size[2], x_size[3]])
+    x = tf.slice(x, [1, 0, 0, 0], [-1, -1, -1, -1])
+    x = tf.reshape(x, [x_size[0], x_size[1] - 1, x_size[2], x_size[3]])
+    x = tf.slice(x, [0, 0, 0, 0], [-1, klen, -1, -1])
+
+    x = tf.transpose(x, perm=[2, 3, 0, 1])
+
+    return x
+
+
+@keras_nlp_export("keras_nlp.layers.MultiHeadRelativeAttention")
+class MultiHeadRelativeAttention(keras.layers.MultiHeadAttention):
+    """A multi-head attention layer with relative attention + position encoding.
+    This layer shares the same input/output projections as the common
+    `keras.layers.MultiHeadAttention` layer.
+    When it calculates attention logits, position encoding is projected to form
+    relative keys. The logits are composed by shifted relative logits and content
+    logits.
+    **Note: This layer is currently experimental.
+    Attributes:
+      kernel_initializer: The kernel initializer. Defaults to variance_scaling.
+    Call args:
+      query: Query `Tensor` of shape `[B, T, dim]`.
+      value: Value `Tensor` of shape `[B, S, dim]`.
+      content_attention_bias: Bias `Tensor` for content based attention of shape
+        `[num_heads, dim]`.
+      positional_attention_bias: Bias `Tensor` for position based attention of
+        shape `[num_heads, dim]`.
+      key: Optional key `Tensor` of shape `[B, S, dim]`. If not given, will use
+        `value` for both `key` and `value`, which is the most common case.
+      relative_position_encoding: Relative positional encoding `Tensor` of shape
+        `[B, L, dim]`.
+      segment_matrix: Optional `Tensor` representing segmentation IDs used in
+        XLNet of shape `[B, S, S + M]`.
+      segment_encoding: Optional `Tensor` representing the segmentation encoding
+        as used in XLNet of shape `[2, num_heads, dim]`.
+      segment_attention_bias: Optional trainable bias parameter added to the query
+        had when calculating the segment-based attention score used in XLNet of
+        shape `[num_heads, dim]`.
+      state: Optional `Tensor` of shape `[B, M, E]` where M is the length of the
+        state or memory. If passed, this is also attended over as in Transformer
+        XL.
+      attention_mask: A boolean mask of shape `[B, T, S]` that prevents attention
+        to certain positions.
+    """
+
+    def __init__(self, kernel_initializer="variance_scaling", **kwargs):
+        super().__init__(kernel_initializer=kernel_initializer, **kwargs)
+
+    def _build_from_signature(self, query, value, key=None):
```

Comment:
If there is a lot of overlap here with the superclass' `_build_from_signature`, we should just chain to `super()._build_from_signature()`. Let's avoid code duplication as much as possible!